multisig-transactions-monitor/src/agent.js

```diff
@@ -0,0 +1,128 @@
+const { ethers } = require('forta-agent');
+
+// load agent configuration parameters
+const config = require('../agent-config.json');
+
+// require utilities
+const utils = require('./utils');
+
+// set up variable to hold initalization data for use in the handler
+const initializeData = {};
+
+function provideInitialize(data) {
+  return async function initialize() {
+    /* eslint-disable no-param-reassign */
+
+    // get protocol and developer info
+    data.protocolName = config.protocolName;
+    data.protocolAbbreviation = config.protocolAbbreviation;
+    data.developerAbbreviation = config.developerAbbreviation;
+
+    const sigTypeFull = ethers.utils.FormatTypes.full;
+
+    // get contracts' abi and monitored event signatures
+    const { contracts } = config;
+    data.contractsInfo = Object.entries(contracts).map(([name, entry]) => {
+      const { events: eventNames, abiFile } = entry;
+      const file = utils.getAbi(abiFile);
+      const contractInterface = new ethers.utils.Interface(file.abi);
+      const eventSignatures = eventNames.map((eventName) => {
+        const signature = contractInterface.getEvent(eventName).format(sigTypeFull);
+        return signature;
+      });
+
+      // create object to store and return necessary contract information
+      const contract = {
+        name,
+        eventSignatures,
+        eventNames,
+      };
+
+      return contract;
+    });
+
+    // get contrat addresses
+    data.multisigAddress = (config.contracts.multisig.address).toLowerCase();
+    data.governanceAddress = (config.contracts.governance.address).toLowerCase();
+    data.comptrollerAddress = (config.contracts.comptroller.address).toLowerCase();
+  };
+}
+
+function provideHandleTransaction(data) {
+  return async function handleTransaction(txEvent) {
+    const {
+      protocolName,
+      protocolAbbreviation,
+      developerAbbreviation,
+      contractsInfo,
+      multisigAddress,
+      governanceAddress,
+      comptrollerAddress,
+    } = data;
+
+    const findings = [];
+
+    // filter for transactions involving the multisig address
+    const txAddrs = Object.keys(txEvent.addresses).map((address) => address.toLowerCase());
+
+    // if the multisig was involed in a transaction, find out specifically which one
+    if (txAddrs.indexOf(multisigAddress) !== -1) {
+      contractsInfo.forEach((contract) => {
+        // filter for which event and address the multisig transaction was involved in
+        const parsedLogsMultiSig = txEvent.filterLog(contract.eventSignatures, multisigAddress);
+        const parsedLogsGovernance = txEvent.filterLog(contract.eventSignatures, governanceAddress);
+        const parsedLogsComptroller = txEvent.filterLog(
+          contract.eventSignatures,
+          comptrollerAddress,
+        );
```

Comment:
On first blush, this looked like it would be okay, but after looking at it more, I expect that the following behavior will be seen, based upon the ordering of the entries in `agent-config.json`:
- First pass, `parsedLogsMultisig` may be non-empty, but the other two will be empty Arrays
- Second pass, `parsedLogsGovernance` may be non-empty, but the other two will be empty Arrays
- Third pass, `parsedLogsComptroller` may be non-empty, but the other two will be empty Arrays

Under the hood, `filterLog` is going to perform the following:
- Filter logs based on the matching address
- Create an ethers.js Interface based upon the eventSignatures passed in
- Use the Interface.parseLog method to match log topic hashes with what was passed in the eventSignatures

For the first iteration, the values in `contract.eventSignatures` will correspond to the event signatures for the multisig contract.  Those signatures don't match any events in the Governance or Comptroller contracts, so `filterLog` should always return empty for those.  The same pattern repeats for iterations two and three.

I suggest the following:
- Store the contract addresses in the `contract` Objects that have the `name`, `eventNames`, and `eventSignatures` in them.
- Also store the appropriate `utils.createXYZFinding` function Object in the `contract` Objects for the respective contracts.
- Keep another copy of the multisig address in the original place (i.e. `data.multisigAddress`) so you can perform the address filtering before doing any `filterLog` checks

With those changes, this entire section will simplify to something like:
```javascript
      contractsInfo.forEach((contract) => {
        // filter for which event and address the multisig transaction was involved in
        const parsedLogs = txEvent.filterLog(contract.eventSignatures, contract.address);

        if (parsedLogs.length !== 0) {
          parsedLogs.forEach((log) => {
            const finding = contract.createFindingFunction(
              log,
              protocolName,
              protocolAbbreviation,
              developerAbbreviation,
            );
            findings.push(finding);
          });
        }
      }
```

Reply:
good catch and thanks for the explanation. Will change it to this